external/storm-kafka/src/jvm/storm/kafka/bolt/KafkaBolt.java

```diff
@@ -102,12 +114,40 @@ public void execute(Tuple input) {
             key = mapper.getKeyFromTuple(input);
             message = mapper.getMessageFromTuple(input);
             topic = topicSelector.getTopic(input);
-            if(topic != null ) {
-                producer.send(new KeyedMessage<K, V>(topic, key, message));
+            if (topic != null ) {
+                Callback callback = null;
+
+                if (!fireAndForget && async) {
+                    callback = new Callback() {
+                        @Override
+                        public void onCompletion(RecordMetadata ignored, Exception e) {
+                            synchronized (collector) {
```

Comment:
Most of the storm user facing components usually thought as a single thread actors. Are you sure you want to break this abstraction?